AndroidTest/app/src/main/java/cloudant/com/androidtest/keyprovider/AndroidKeyProviderTests.java

```diff
@@ -0,0 +1,173 @@
+                ProviderTestUtil.getUniqueIdentifier());
+
+        EncryptionKey createdKey = provider.getEncryptionKey();
+        assertNotNull("EncryptionKey was not created", createdKey);
+
+        EncryptionKey loadedKey = provider.getEncryptionKey();
+        assertNotNull("EncryptionKey was not loaded", loadedKey);
+
+        assertTrue("loadedKey did not match createdKey", Arrays.equals(createdKey.getKey(),
+                loadedKey.getKey()));
+    }
+
+    public void testCreateProviderWithManager() {
+        KeyStorage storage = new KeyStorage(getContext(), ProviderTestUtil.getUniqueIdentifier());
+        KeyManager manager = new KeyManager(storage);
+        AndroidKeyProvider provider = new AndroidKeyProvider("provider1password", manager);
+
+        EncryptionKey createdKey = provider.getEncryptionKey();
+        assertNotNull("EncryptionKey was not created", createdKey);
+
+        EncryptionKey loadedKey = provider.getEncryptionKey();
+        assertNotNull("EncryptionKey was not loaded", loadedKey);
+
+        assertTrue("loadedKey did not match createdKey", Arrays.equals(createdKey.getKey(),
+                loadedKey.getKey()));
+    }
+
+    public void testCreateTwoProvidersWithSameIdentifier() {
+        String identifier = ProviderTestUtil.getUniqueIdentifier();
+        AndroidKeyProvider provider1 = new AndroidKeyProvider(getContext(), "provider1password",
+                identifier);
+        AndroidKeyProvider provider2 = new AndroidKeyProvider(getContext(), "provider1password",
+                identifier);
+
+        EncryptionKey provider1Key = provider1.getEncryptionKey();
+        assertNotNull("provider1 EncryptionKey was not created", provider1Key);
+
+        EncryptionKey provider2Key = provider2.getEncryptionKey();
+        assertNotNull("provider2 EncryptionKey was not created", provider2Key);
+
+        assertTrue("provider1 key and provider2 key should be equal", Arrays.equals(provider1Key
+                .getKey(), provider2Key.getKey()));
+    }
+
+    public void testCreateTwoProvidersWithDifferentIdentifier() {
+        AndroidKeyProvider provider1 = new AndroidKeyProvider(getContext(), "provider1password",
+                ProviderTestUtil.getUniqueIdentifier());
+        AndroidKeyProvider provider2 = new AndroidKeyProvider(getContext(), "provider2password",
+                ProviderTestUtil.getUniqueIdentifier());
+
+        EncryptionKey provider1Key = provider1.getEncryptionKey();
+        assertNotNull("provider1 EncryptionKey was not created", provider1Key);
+
+        EncryptionKey provider2Key = provider2.getEncryptionKey();
+        assertNotNull("provider2 EncryptionKey was not created", provider2Key);
+
+        assertFalse("provider1 key and provider2 key should not be equal", Arrays.equals
+                (provider1Key.getKey(),
+                        provider2Key.getKey()));
+    }
+
+    // Negative tests
+    public void testCreateProviderWithNullContext() {
```

Comment:
These sorts of tests can be rewritten as
```
    @Test(expected = IllegalArgumentException.class)
```
Although it's not everyone's favourite idiom and it doesn't allow you to make any other assertions or introspect on the detail of the exception thrown. Personally I also wouldn't bother catching other exceptions and failing, just let JUnit do its job in that respect.